Feat: Add redis backend

p1gp1g opened the pull request:
Currently, autopush requires using Bigtable, there is no particular issue for a big, public server, but some may want to host their own server, or to do some tests locally. So this PR adds another kind of backend using Redis. It should work with redis forks too.

It mostly follow the behavior of bigtable with a few differences:
- when saving a message with a topic, it simply removes the previous message for this topic. So topic messages and _timestamp_ messages are processed the exact same way. Therefore the test (`run_gauntlet`) is a bit different for this: it adds 2 messages for the same topic and check only one is fetched.
- Timestamps used and returned by `fetch_timestamp_messages` are the date of the record in ms, with bigtables it uses the message's `sortkey_timestamp`, in seconds.

I've also added a docker-compose file to run everything.

For the moment _RedisClientImpl_ doesn't use connection pooling, but this should be easily implemented, it also provide the Command trait, returned by `RedisClientImpl.connection`

Fixes #774

jrconlin commented on autopush-common/src/notification.rs:
```diff
@@ -47,6 +47,9 @@ impl Notification {
+        #[cfg(feature = "redis")]
+        return format!("{}:{}", chid, self.version);
+
         if let Some(ref topic) = self.topic {
```
I'd be surprised if clippy doesn't complain about this. 
Might want to add a feature flag before this line for `#[cfg(feature = "bigtable")] or I guess `#[cfg(not(feature = "redis"))]`

Honestly, you could probably use this untouched for the redis key to make the Topic messages work correctly.

p1gp1g replied:
It works well without touching this function. It actually introduced a big if it was compiled with the 2 features

jrconlin commented on docker-compose.yml:
```diff
@@ -0,0 +1,51 @@
+services:
```
Let's rename this as `redis-docker-compose.yml` so that it's clear that this should only be used for redis deployments.

jrconlin reviewed (changes requested):
Gotta admit, Redis is not really the DB that I would have picked for this, but sure, provided that the durable storage element is set, it should be able to handle light loads. 

FWIW, I had done a version that used Postgres a long time ago. (If you go through the Changelog, you might find a reference to it.) Sadly, I burned that branch a while ago.

Still, thank you for this. There are a few things you'll need to do to fix it up.

The requirements are:
* Please don't use `.unwrap()` (or if you do, please explain why.)
* We actually do need to support topic messages.
* (Make sure that all unit and integration tests pass)

p1gp1g commented on autopush-common/src/db/redis/redis_client/mod.rs:
```diff
@@ -0,0 +1,842 @@
+            .ok_or(DbError::General("Could not find DSN".to_owned()))?;
+        let client = redis::Client::open(dsn).unwrap();
+        let db_settings = RedisDbSettings::try_from(settings.db_settings.as_ref())?;
+        info!("🉑 {:#?}", db_settings);
```
Done with the 🐰 :+1:

p1gp1g commented on autopush-common/build.rs:
```diff
@@ -1,5 +1,5 @@
 pub fn main() {
-    if !cfg!(feature = "bigtable") {
-        panic!("No database defined! Please compile with `features=bigtable`");
+    if !cfg!(feature = "bigtable") && !cfg!(feature = "redis") {
```
They should not be conflicting, the server is selected with the dsn url

p1gp1g replied:
I agree there is no use case when both are used, but I can see some examples for compiling both. For instance if a generic container image is published/used, or one do a build that can be used in different environments, (dev/tests with redis, integration/prod with bigtable)

p1gp1g commented on autopush-common/src/db/redis/redis_client/mod.rs:
```diff
@@ -0,0 +1,842 @@
+        let mut con = self.connection()?;
+        let user_key = self.user_key(&user.uaid);
+        let co_key = self.last_co_key(&user.uaid);
+        let _: () = redis::pipe()
```
I've moved to multiplexed connection

p1gp1g commented:
Thank you for the review, I've addressed most of the comments.

I went to redis, because this is also what I used for [Uppush](https://codeberg.org/NextPush/uppush), which also work as a webpush server. I've seen in the issue comments you had a version with Postgres.

Regarding the topics, this implementation supports topic, but in another way. See the tests: we store 2 messages with the same topic and fetch a single one

```rs
        // We store 2 messages, with a single topic
        let test_notification_0 = crate::db::Notification {
            channel_id: topic_chid,
            version: "version0".to_owned(),
            ttl: 300,
            topic: Some("topic".to_owned()),
            timestamp,
            data: Some(test_data.clone()),
            sortkey_timestamp: Some(sort_key),
            ..Default::default()
        };
        assert!(client
            .save_message(&uaid, test_notification_0.clone())
            .await
            .is_ok());

        let test_notification = crate::db::Notification {
            timestamp: now(),
            version: "version1".to_owned(),
            sortkey_timestamp: Some(sort_key + 10),
            ..test_notification_0
        };

        assert!(client
            .save_message(&uaid, test_notification.clone())
            .await
            .is_ok());

        let mut fetched = client.fetch_timestamp_messages(&uaid, None, 999).await?;
        assert_eq!(fetched.messages.len(), 1);
```

By the way, using the current `chidmessageid` allows to do this automatically, the topic isn't necessary: https://github.com/mozilla-services/autopush-rs/pull/813/commits/ddc8a465d6840ac4253d5f4501168d46be41b8c5

Regarding the tests, they all pass `cargo test --no-default-features --features redis`

p1gp1g commented on autopush-common/src/db/redis/redis_client/mod.rs:
```diff
@@ -0,0 +1,842 @@
+use super::RedisDbSettings;
+
+/// Semi convenience wrapper to ensure that the UAID is formatted and displayed consistently.
+// TODO:Should we create something similar for ChannelID?
```
It also avoid passing a channel id for a uaid by mistake

p1gp1g commented on autopush-common/src/db/redis/redis_client/mod.rs:
```diff
@@ -0,0 +1,842 @@
+    /**
+    Topic messages are handled as other messages with redis, we return nothing.
+    */
+    async fn fetch_topic_messages(
```
Topic messages are replaced and only the last version is kept. See comment bellow

jrconlin commented:
Sigh, just like the other PR, I'm afraid [we can't accept this PR with unsigned commits](https://github.com/mozilla-services/autopush-rs/blob/master/CONTRIBUTING.md#sending-pull-requests). Please feel free to resubmit this with [signed commits](https://github.com/mozilla-services/autopush-rs/blob/master/CONTRIBUTING.md#sending-pull-requests). 

(This is a policy put in place by our Security Operations team.)

p1gp1g commented:
All the commits are signed

jrconlin commented on autopush-common/src/db/redis/redis_client/mod.rs:
```diff
@@ -0,0 +1,855 @@
+    ///
+    /// If [`limit`] = 0, we fetch all messages after [`timestamp`].
+    ///
+    /// This can return expired messages, following bigtables behavior
```
Yeah, but there's no real need to do that, and in fact, it's probably an anti-pattern we need to fix in bigtable. (There's a potential abuse vector that should be cut off.)

Ideally, messages should be dropped once they've been decisively handled (e.g. if they're ACK'd or if the NACK response specifies that the message is undeliverable. see https://github.com/mozilla-services/autopush-rs/pull/795 )

p1gp1g replied:
Expired messages are removed from redis, so it doesn't use memory for expired messages.

But it has to return dummy expired events to avoid `check_storage_advance` to stop [if there are 10 expired messages](https://github.com/mozilla-services/autopush-rs/blob/a6fc8cdb678e4bc538e68bab601f801ef8c4b2af/autoconnect/autoconnect-ws/autoconnect-ws-sm/src/identified/on_server_notif.rs#L109) (so messages is an empty vec)

> Ideally, messages should be dropped once they've been decisively handled (e.g. if they're ACK'd or if the NACK response specifies that the message is undeliverable. see https://github.com/mozilla-services/autopush-rs/pull/795 )

`remove_message` do remove the message, so this is already the case :)

jrconlin commented on autopush-common/src/db/redis/redis_client/mod.rs:
```diff
@@ -0,0 +1,855 @@
+                .into_iter()
+                .filter_map(|opt: Option<String>| {
+                    if opt.is_none() {
+                        // We return dummy expired event if we can't fetch the say event,
```
```suggestion
                        // We return dummy expired event if we can't fetch the said event,
```

p1gp1g commented on autopush-common/src/db/redis/redis_client/mod.rs:
```diff
@@ -0,0 +1,855 @@
+    }
+
+    /**
+    Return a [ConnectionLike], which implement redis [Commands] and can be
```
This is a block doctstring `/**`, I'm using it in other projects, habit from kotlin. I can move to line docstring :+1:

jrconlin replied:
Ah, I figured you might want it included if anyone were to run `rustdoc`. Not a big issue, just commenting.

p1gp1g replied:
Fixed

p1gp1g commented:
I have rebased this branch too

jrconlin commented:
Sigh. Just noticed that these commits aren't signed. [We require all commits to be signed](https://github.com/mozilla-services/autopush-rs/blob/master/CONTRIBUTING.md#sending-pull-requests) before we can accept them. (I've tried to make that very clear in the docs and PR templates, but I guess there's still work to do)

Sadly, that means we can't accept this PR as is (even one unsigned commit corrupts things.)

You can either close this PR and resubmit it or I can submit it on your behalf, but I'll need one of the other core folk to sign off on it.

p1gp1g commented:
I have squashed the branch and signed the commit, let me know if you prefer keep the different commits for the review

p1gp1g commented:
That'd be awesome ! Let me know when you have more information

p1gp1g commented:
@jrconlin: I think we can close this PR now #1039 is merged? And thanks for having polished, and merged the feature :)
